src/tree/updater_quantile_hist.cc

```diff
@@ -95,43 +104,151 @@ bool QuantileHistMaker::UpdatePredictionCache(
+
+    if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1) {
+      const size_t parent_id = (*p_tree)[entry.nid].Parent();
+      auto parent_hist = builder->hist_local_worker_[parent_id];
+      auto sibling_hist = builder->hist_[entry.sibling_nid];
       SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
+      // Store posible parent node
+      auto sibling_local = builder->hist_local_worker_[entry.sibling_nid];
+      CopyHist(sibling_local, sibling_hist, r.begin(), r.end());
     }
   });
+  builder->builder_monitor_.Start("SyncHistogramsAllreduce");
+  builder->histred_.Allreduce(builder->hist_[starting_index].data(),
+                                    builder->hist_builder_.GetNumBins() * sync_count);
+  builder->builder_monitor_.Stop("SyncHistogramsAllreduce");
+
+  ParallelSubtractionHist(builder, space, builder->nodes_for_explicit_hist_build_, p_tree);
+
+  common::BlockedSpace2d space2(builder->nodes_for_subtraction_trick_.size(), [&](size_t node) {
+    return nbins;
+  }, 1024);
+  ParallelSubtractionHist(builder, space2, builder->nodes_for_subtraction_trick_, p_tree);
+  builder->builder_monitor_.Stop("SyncHistograms");
+}
 
-  if (isDistributed) {
-    this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
-    // use Subtraction Trick
-    for (auto const& node : nodes_for_subtraction_trick_) {
-      SubtractionTrick(hist_[node.nid], hist_[node.sibling_nid],
-                       hist_[(*p_tree)[node.nid].Parent()]);
+void DistributedHistSynchronizer::ParallelSubtractionHist(QuantileHistMaker::Builder* builder,
+                                  const common::BlockedSpace2d& space,
+                                  const std::vector<QuantileHistMaker::Builder::ExpandEntry>& nodes,
+                                  const RegTree * p_tree) {
+  common::ParallelFor2d(space, builder->nthread_, [&](size_t node, common::Range1d r) {
+    const auto entry = nodes[node];
+    if (!((*p_tree)[entry.nid].IsLeftChild())) {
+      auto this_hist = builder->hist_[entry.nid];
+
+      if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1) {
+        auto parent_hist = builder->hist_[(*p_tree)[entry.nid].Parent()];
+        auto sibling_hist = builder->hist_[entry.sibling_nid];
+        SubtractionHist(this_hist, parent_hist, sibling_hist, r.begin(), r.end());
+      }
     }
+  });
+}
+
+void BatchHistRowsAdder::AddHistRows(QuantileHistMaker::Builder* builder,
+                                     int *starting_index, int *sync_count,
+                                     RegTree *p_tree) {
+  builder->builder_monitor_.Start("AddHistRows");
+
+  for (auto const& entry : builder->nodes_for_explicit_hist_build_) {
+    int nid = entry.nid;
+    builder->hist_.AddHistRow(nid);
+    (*starting_index) = std::min(nid, (*starting_index));
   }
+  (*sync_count) = builder->nodes_for_explicit_hist_build_.size();
```

Comment:
Minor comment: there is no need to initialize ```starting_index``` and ```sync_count``` by special value in cases of single-node mode. Since it's used only in distributed mode.

Reply:
fixed